feat: add vLLM as a supported runtime

yashaswipiplani opened the pull request:
**What this PR does / why we need it**:
Added vLLM as a separate supported runtime. This is the first PR in separating vLLM and HuggingFace transformers runtimes. Will raise a follow up PR for removing vLLM code from HuggingFaceServer. Created issue to track the same: https://github.com/kserve/kserve/issues/4770

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

**Single Node GPU Runtime**
ISVC applied:
```
kubectl apply -n kserve-test -f - <<EOF
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "qwen-llm"
  namespace: kserve-test
spec:
  predictor:
    model:
      modelFormat:
        name: vllm
      args:
        - --served-model-name=qwen
      storageUri: "hf://Qwen/Qwen2.5-0.5B-Instruct"
      resources:
        limits:
          cpu: "2"
          memory: 6Gi
          nvidia.com/gpu: "1"
        requests:
          cpu: "1"
          memory: 4Gi
          nvidia.com/gpu: "1"
EOF
```
**Single Node CPU Runtime**
ISVC applied:
```
kubectl apply -n kserve-test -f - <<EOF
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "qwen-llm"
  namespace: kserve-test
spec:
  predictor:
    model:
      modelFormat:
        name: vllm
      args:
        - --served-model-name=qwen
        - --max-model-len=4096
      storageUri: "hf://Qwen/Qwen2.5-0.5B-Instruct"
      resources:
        limits:
          cpu: "16"
          memory: 20Gi
        requests:
          cpu: "16"
          memory: 20Gi
EOF
```

**Inference tested for both:**
<img width="1057" height="61" alt="Screenshot 2025-10-22 at 1 59 37 PM" src="https://github.com/user-attachments/assets/865000be-bd9d-4ef4-979e-1425421220c6" />
<img width="1070" height="230" alt="Screenshot 2025-10-22 at 1 59 49 PM" src="https://github.com/user-attachments/assets/73f0d9ee-c6d4-41df-b21e-0cc60aa369f7" />

**Multi node inferencing**
ISVC Applied:
```
kubectl apply -n kserve-test -f - <<EOF
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "qwen-llm"
  namespace: kserve-test
  annotations:
    serving.kserve.io/autoscalerClass: "none"
spec:
  predictor:
    model:
      modelFormat:
        name: vllm
      args:
        - --served-model-name=qwen
      storageUri: "pvc://qwen-model-pvc/Qwen2.5-0.5B-Instruct"
      resources:
        limits:
          cpu: "2"
          memory: 6Gi
          nvidia.com/gpu: "1"
        requests:
          cpu: "1"
          memory: 4Gi
          nvidia.com/gpu: "1"
    workerSpec:
      tensorParallelSize: 1
      pipelineParallelSize: 2
      resources:
        limits:
          cpu: "2"
          memory: 6Gi
          nvidia.com/gpu: "1"
        requests:
          cpu: "1"
          memory: 4Gi
          nvidia.com/gpu: "1"
EOF
```

**Inference testing:**
<img width="1072" height="123" alt="Screenshot 2025-10-19 at 1 15 15 PM" src="https://github.com/user-attachments/assets/4bab02d9-0a7f-4de6-9a34-83d768be22f0" />
<img width="1061" height="201" alt="Screenshot 2025-10-19 at 1 15 36 PM" src="https://github.com/user-attachments/assets/4502dec6-986a-4f12-9051-e9124596ccc7" />


**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

gavrissh commented:
sign-off your PR with these steps - https://github.com/kserve/kserve/pull/4769/checks?check_run_id=53384248700

gavrissh commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -708,3 +708,323 @@ spec:
+      autoSelect: true
+      priority: 1
+  protocolVersions:
+    - v1
```
vLLM runtime will not support v1 protocol

yashaswipiplani replied:
Removed

gavrissh commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -708,3 +708,323 @@ spec:
+      env:
+        - name: LMCACHE_USE_EXPERIMENTAL
+          value: "{{ .Values.kserve.servingruntime.vllmserver.lmcacheUseExperimental }}"
+        - name: VLLM_GPU_IMAGE
```
Could you describe what this change does in the description

yashaswipiplani replied:
Added the description, please review now. Thanks

gavrissh commented on test/e2e/predictor/test_vllm.py:
```diff
@@ -0,0 +1,745 @@
+                "bfloat16",
+            ],
+            resources=V1ResourceRequirements(
+                requests={"cpu": "2", "memory": "8Gi", "nvidia.com/gpu": "1"},
```
There are no GPUs in GitHub workers; we can run only on CPU tests

yashaswipiplani replied:
Removed GPU

spolti commented on pkg/controller/v1beta1/inferenceservice/utils/utils.go:
```diff
@@ -378,6 +378,11 @@ func UpdateImageTag(container *corev1.Container, runtimeVersion *string, serving
+			} else if servingRuntime != nil && *servingRuntime == constants.VLLMServer {
+				// For vLLM server, use the GPU image from environment variable if GPU Enabled
+				if image, exists := utils.GetEnvVarValue(container.Env, "VLLM_GPU_IMAGE"); exists {
+					container.Image = image
```
I am not sure if it is a good approach; ideally, we need to have VLLM CPU and GPU-specific runtimes, since the parameters and configurations aren't the same.

yashaswipiplani replied:
I was thinking about a constant vLLM GPU image which we can use to override image in controller if GPU is enabled. This seems like a much simpler approach and close to what we already have for HuggingFaceServer. Do you think we should separate out vLLM CPU and GPU runtimes? Because, the controller can't select GPU runtime if GPU is enabled as of now. 
Also, the parameters and configuration remains almost the same for single node GPU and CPU runtime, right? The difference is only in the case of multi-node

spolti replied:
For vLLM I think we should treat it differently, and not rely on the env as you did.
Essentially a similar approach to multi-node to detect GPUs, if no GPU is available, fallback to the cpu vllm container.

spolti commented on test/e2e/predictor/test_vllm.py:
```diff
@@ -0,0 +1,745 @@
+            env=[
+                client.V1EnvVar(
+                    name="MODEL_ID",
+                    value="Qwen/Qwen2-0.5B-Instruct",
```
This model size is around 1.2 GB. Having to download it for every test is not a good thing. Can we use the tiniest model available for VLLM?

spolti commented on config/runtimes/kserve-vllmserver.yaml:
```diff
@@ -0,0 +1,81 @@
+  containers:
+    - name: kserve-container
+      image: vllmserver:replace
+      command:
```
Where are the args?
You can use this as an example https://github.com/opendatahub-io/odh-model-controller/blob/incubating/config/runtimes/vllm-cpu-template.yaml

And provide gpu and cpu runtimes independently.

yashaswipiplani replied:
`vllm serve ${MODEL} --host 0.0.0.0 --port 8000 $@`
Arguments that are provided in isvc are appended to the vLLM serve command itself. I was thinking we do not add any default arguments and let the users add whatever arguments they want?

spolti replied:
having default is needed so user does not need to worry about it every time.
also, please see my comments [here](https://github.com/kserve/kserve/pull/4769/changes#diff-9ff2b26b73b2bedec0ae977c62090223372b509bed0ed577adc91088b1f8fa4b) and apply as needed.

cjackal commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -708,3 +708,321 @@ spec:
+          echo "Ray head node is ready"
+
+          # Start vLLM Server with multi-node configuration
+          vllm serve ${MODEL} --host 0.0.0.0 --port 8000 --tensor-parallel-size ${TENSOR_PARALLEL_SIZE:-1} --pipeline-parallel-size ${PIPELINE_PARALLEL_SIZE:-1} $@
```
Don't we need to wait for the worker pods to be up and joined? huggingface runtime has [this check](https://github.com/kserve/kserve/blob/00e88c51ff62455d5a22fca53cdeadc3e14a53d6/python/huggingfaceserver/health_check.py#L117-L152) for example.

yashaswipiplani replied:
Added the check in the pod. Please check now. Thanks

gavrissh commented:
/rerun-all

sivanantha321 commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -708,3 +708,347 @@ spec:
+    prometheus.kserve.io/port: '8000'
+    prometheus.kserve.io/path: "/metrics"
+  supportedModelFormats:
+    - name: vllm
```
Not sure if it is right to name the model format as vllm.

yashaswipiplani replied:
Can you please suggest alternate appropriate name for the same. Thanks

sivanantha321 commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -708,3 +708,347 @@ spec:
+apiVersion: serving.kserve.io/v1alpha1
+kind: ClusterServingRuntime
+metadata:
+  name: kserve-vllmserver-multinode
```
isn't it better to use llminferenceservice for multi node ?

yashaswipiplani replied:
Removed from clusterservingruntimes

yashaswipiplani replied:
Can you please review the changes again. Thanks

gavrissh commented:
/rerun-all

Datta0 commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -708,3 +708,347 @@ spec:
+          done
+
+          # Start vLLM Server with multi-node configuration
+          vllm serve ${MODEL} --host 0.0.0.0 --port 8000 --tensor-parallel-size ${TENSOR_PARALLEL_SIZE:-1} --pipeline-parallel-size ${PIPELINE_PARALLEL_SIZE:-1} $@
```
Not really in favor of having TP and PP sizes in the args here. Shouldn't we let the user decide what config they want? Also, iirc $@ passes all the args entered at the start of the command. So not sure what happens if someone specifies --tensor-parallel-size in the runtime args again?

yashaswipiplani replied:
This is removed from clusterservingruntimes altogether now. We will use llminferenceservice for multi node vLLM

Datta0 commented on charts/kserve-resources/values.yaml:
```diff
@@ -361,6 +361,36 @@ kserve:
         enabled: true
         sizeLimit: "3Gi"
+    vllmserver:
+      disabled: false
```
Is there a plan to bake in other vLLM CPU images like with AMX/open vino optimisations? Or is this supposed to be just a basic vllm cpu image with no updates scoped ?

yashaswipiplani replied:
Not sure about this. Probably @gavrissh or @johnugeorge can answer this

gavrissh replied:
The current CPU image here is an AMX-enabled one. 
OpenVino is deprecated in vLLM

Datta0 commented on pkg/apis/serving/v1beta1/inference_service_defaults.go:
```diff
@@ -244,8 +244,8 @@ func (isvc *InferenceService) setPredictorModelDefaults() {
 	if isvc.Spec.Predictor.Model != nil && isvc.Spec.Predictor.Model.ProtocolVersion == nil {
-		if isvc.Spec.Predictor.Model.ModelFormat.Name == constants.SupportedModelTriton {
-			// set 'v2' as default protocol version for triton server
+		if isvc.Spec.Predictor.Model.ModelFormat.Name == constants.SupportedModelTriton || isvc.Spec.Predictor.Model.ModelFormat.Name == constants.SupportedModelVLLM {
```
Can you describe/explain this change?

yashaswipiplani replied:
Yes, so vLLM runtime will not support V1 protocol. Now when we apply an ISVC, the controller tries to use the V1 protocol and fails since it is not supported in runtime. Defaulting it to use V2 protocol, similar to Triton server. Without this change, we will need to specify to use V2 protocol in ISVC

gavrissh commented:
/rerun-all

sivanantha321 commented:
@yuzisun please provide your thoughts on this

terrytangyuan reviewed (approved):
+1 for a separate runtime. Although I am not sure if we should change the existing HF runtime since it may impact existing users. 

cc @Jooho @yuzisun

spolti commented:
+1 to keep it separate.

@yashaswipiplani can you please help with fixing the conflicts?

spolti commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -720,3 +720,101 @@ spec:
+    prometheus.kserve.io/port: '8000'
+    prometheus.kserve.io/path: "/metrics"
+  supportedModelFormats:
+    - name: vllm
```
```suggestion
    - name: vLLM
```

spolti commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -720,3 +720,101 @@ spec:
+      version: "1"
+      autoSelect: true
+      priority: 1
+  protocolVersions:
```
I think we can omit it.

Jooho replied:
That’s correct—since we’re using the OpenAI API with the vLLM model, protocolVersion doesn’t seem relevant in this context.

spolti commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -720,3 +720,101 @@ spec:
+  containers:
+    - name: kserve-container
+      image: "{{ .Values.kserve.servingruntime.vllmserver.image }}:{{ .Values.kserve.servingruntime.vllmserver.tag }}"
+      command:
```
the command can be as simple as:

```yaml
          command:
            - python
            - -m
            - vllm.entrypoints.openai.api_server
          args:
            - "--port=8080"
            - "--model=/mnt/models"
            - "--served-model-name={{.Name}}"
```

spolti commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -720,3 +720,101 @@ spec:
+      env:
+        - name: LMCACHE_USE_EXPERIMENTAL
+          value: "{{ .Values.kserve.servingruntime.vllmserver.lmcacheUseExperimental }}"
+        - name: VLLM_GPU_IMAGE
```
What is the purpose of it?

spolti commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
```diff
@@ -720,3 +720,101 @@ spec:
+          
+          vllm serve ${MODEL} --host 0.0.0.0 --port 8000 $@
+        - "bash"
+      env:
```
We would also need to set HF_HOME to some place where the storageInitializer has write permissions on it to ensure it will run fine on more restricted environments.

spolti commented:
also, I guess you need a full rebase, there are other commits on this PR.

Jooho commented on config/runtimes/kserve-vllmserver.yaml:
```diff
@@ -0,0 +1,85 @@
+      autoSelect: true
+      priority: 1
+  protocolVersions:
+    - v2
```
since we’re using the OpenAI API with the vLLM model, protocolVersion doesn’t seem relevant in this context.

Jooho commented on config/runtimes/kserve-vllmserver.yaml:
```diff
@@ -0,0 +1,85 @@
+    prometheus.kserve.io/port: "8000"
+    prometheus.kserve.io/path: "/metrics"
+  supportedModelFormats:
+    - name: vllm
```
I’m fine with vllm, but since vLLM is the most commonly used form in the documentation, I think vLLM would be preferable.

Jooho commented on charts/kserve-resources/templates/clusterservingruntimes.yaml:
This YAML file will be generated automatically based on kustomize yaml file so do not update any in this file. Instead, please update the changes on kustomize yaml.

Jooho commented on config/runtimes/kserve-vllmserver.yaml:
If we have a ServingRuntime for CPU, shouldn’t we also have ServingRuntimes for other accelerators such as GPU/rocm? If so we need to change this file/sr name to `kserve-vllmserver-cpu.yaml`

Jooho commented on config/runtimes/kserve-vllmserver.yaml:
```diff
@@ -0,0 +1,85 @@
+          name: http
+          protocol: TCP
+      env:
+        - name: LMCACHE_USE_EXPERIMENTAL
+          value: "True"
+        - name: VLLM_GPU_IMAGE
+          value: vllmserver-gpu:replace
```
These env are required?  how this gpu image used? I thought this servingruntime is for vllm cpu.

Jooho commented on config/runtimes/kserve-vllmserver.yaml:
```diff
@@ -0,0 +1,85 @@
+        timeoutSeconds: 15
+      readinessProbe:
+        httpGet:
+          path: /health
```
This does not aware that model is loaded or not.

Jooho commented on config/runtimes/kserve-vllmserver.yaml:
```diff
@@ -0,0 +1,85 @@
+          memory: 2Gi
+      livenessProbe:
+        httpGet:
+          path: /health
```
I think we may need a more refined approach. For liveness, it might be worth requesting a single token, since that would provide a more accurate signal.

Jooho commented on config/runtimes/kserve-vllmserver.yaml:
```diff
@@ -0,0 +1,85 @@
+        periodSeconds: 5
+        successThreshold: 1
+        timeoutSeconds: 15
+      startupProbe:
```
If this is for waiting for loading a model, `/health` would not be the right way.

Jooho commented on config/runtimes/kustomization.yaml:
```diff
@@ -58,4 +59,12 @@ images:
+  newName: public.ecr.aws/q9t5s3a7/vllm-cpu-release-repo
+  newTag: latest
+
+- name: vllmserver-gpu
```
do we have SR for this ?

yashaswipiplani committed:
feat: add vLLM ClusterServingRuntime

Adds a dedicated vLLM ClusterServingRuntime so InferenceService can target
the vllm model format directly, with automatic CPU/GPU image selection.

- Register kserve-vllmserver runtime (config/runtimes + Helm chart patches)
- Add `vllm` SupportedModelFormat constant and default protocol v2
- Override container image with VLLM_GPU_IMAGE when GPU is requested
- Helm values for image, tag, GPU image/tag, securityContext, devShm, hostIPC
- Unit tests for UpdateImageTag covering CPU/GPU/proxy/missing-env cases
- E2E tests for vLLM predictor + classify endpoint helpers

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

yashaswipiplani committed:
Merge branch 'master' into add-vllm-runtime

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Address review comments

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

gavrissh commented:
/rerun-all

yashaswipiplani committed:
Precommit and go tests fix

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

yashaswipiplani committed:
Run make precommit

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Fix e2e tests

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

spolti commented:
/rerun-all

neelabhsinha commented:
/rerun-all

yashaswipiplani committed:
Create separate job for vLLM runtime e2e tests

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Fix timeout and workers for test-vllm-runtime e2e jobs

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

gavrissh commented:
/rerun-all

yashaswipiplani committed:
Fix e2e tests

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

neelabhsinha commented:
/rerun-all

yashaswipiplani committed:
Add buffer between delete and creation of new predictor pod in e2e tests

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

neelabhsinha commented:
/rerun-all

yashaswipiplani committed:
e2e tests fix

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

yashaswipiplani committed:
Correct embedding assertion in e2e tests

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

gavrissh commented:
/rerun-all

yashaswipiplani committed:
Precommit fix

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

gavrissh commented:
/rerun-all

yashaswipiplani committed:
Intermittent e2e failures fix

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

neelabhsinha commented:
/rerun-all

yashaswipiplani committed:
e2e tests fix

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

yashaswipiplani committed:
e2e tests fix

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Increase knative progress deadline for e2e tests

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Address review comments

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Add runAsNonRoot true in security context of vLLM SR

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

yashaswipiplani committed:
Remove runAsNonRoot for vLLM SR

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

Jooho commented:
/rerun-all

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

neelabhsinha commented:
/rerun-all

gavrissh commented:
/rerun-all

neelabhsinha commented:
/rerun-workflow E2E Tests

Jooho commented:
/rerun-all

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

spolti commented:
/rerun-all

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

neelabhsinha commented:
/rerun-all

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

neelabhsinha commented:
/rerun-all

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

neelabhsinha commented:
/rerun-all

gavrissh commented:
/rerun-all

cjohannsen-cloudera commented:
/rerun-all

yashaswipiplani committed:
Empty commit to get ci triggered

Signed-off-by: yashaswipiplani <yashaswi.piplani@nutanix.com>

bartoszmajsak commented:
/rerun-failed

bartoszmajsak commented:
/rerun-failed

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime

yashaswipiplani committed:
Merge branch 'kserve:master' into add-vllm-runtime
